cppapi/server/attrsetval.cpp

```diff
@@ -1457,7 +1457,7 @@ void Attribute::set_value(Tango::DevState *p_data,long x,long y,bool release)
 				*tmp_ptr = *p_data;
 				value.state_seq = new Tango::DevVarStateArray(data_size,data_size,tmp_ptr,release);
 				if (is_fwd_att() == true)
-					delete p_data;
+					delete [] p_data;
 				else
 					delete p_data;
```

Comment:
consider using `SAFE_DELETE(p_data)` from `attribute.h`. There will be redundant check for `release == true` but you will not have duplicate code.

Reply:
Macros are evil and not easy to debug and there was a redundant check. This is indeed why I didn't use the macro in this case.
SAFE_DELETE macro is convenient but it is a nightmare to debug issues in this macro, like https://github.com/tango-controls/cppTango/issues/512 for instance.

Reply:
Maybe we should remove the macro and add something like
* `void safe_delete(T* pointer)` and/or
* `void safe_delete_if_needed(T* pointer, bool needed)`

Also, shall we define a new item for removing all macros? We may start from `server/attribute.h` file.

Reply:
> Maybe we should remove the macro and add something like
> 
>     * `void safe_delete(T* pointer)` and/or
> 
>     * `void safe_delete_if_needed(T* pointer, bool needed)`

I would be in favour of such a change.

> Also, shall we define a new item for removing all macros? We may start from `server/attribute.h` file.

A new issue could be created indeed in order to replace existing macros when possible.

Reply:
I've opened #553 to track this effort.